src/scitoken_helper_fetch.cc

```diff
@@ -22,63 +22,77 @@
   } 
-
-  stringstream default_path;
-  FILE *fruntimedir = GetEnvVarFile("XDG_RUNTIME_DIR", authz_req.pid);
-  string runtimedir;
-  if (fruntimedir != NULL) {
-    GetStringFromFile(fruntimedir, runtimedir);
-    fclose(fruntimedir);
-  }
-  if (runtimedir.size()) {
-    default_path << runtimedir;
-  }
   else {
-    default_path << "/tmp";
-  }
-  default_path << "/bt_u" << authz_req.uid;
-  string default_path_str = default_path.str();
-  if (default_path_str.size() > PATH_MAX) {
-    LogAuthz(kLogAuthzDebug, "default path string bigger than PATH_MAX, ignoring it");
-    default_path_str = "";
+    stringstream default_path;
+    FILE *fruntimedir = GetEnvVarFile("XDG_RUNTIME_DIR", authz_req.pid);
+    string runtimedir;
+    if (fruntimedir != NULL) {
+      GetStringFromFile(fruntimedir, runtimedir);
+      fclose(fruntimedir);
+    }
+    if (runtimedir.size()) {
+      default_path << runtimedir;
+    }
+    else {
+      default_path << "/tmp";
+    }
+    default_path << "/bt_u" << authz_req.uid;
+    string default_path_str = default_path.str();
+    if (default_path_str.size() > PATH_MAX) {
+      LogAuthz(kLogAuthzDebug, "default path string bigger than PATH_MAX, ignoring it");
+      default_path_str = "";
+    }
+
+    ftoken =
+      GetFile(env_name.c_str(), authz_req.pid, authz_req.uid, authz_req.gid, default_path_str);
+    if (ftoken == NULL) {
+      LogAuthz(kLogAuthzDebug, "no token found for %s",
+               authz_req.Ident().c_str());
+      return NULL;
+    }
   }
 
-  FILE *fproxy =
-    GetFile(env_name.c_str(), authz_req.pid, authz_req.uid, authz_req.gid, default_path_str);
-  if (fproxy == NULL) {
-    LogAuthz(kLogAuthzDebug, "no token found for %s",
-             authz_req.Ident().c_str());
-    return NULL;
+  long pos;
+  if ((pos = ftell(ftoken)) == -1) {
+      LogAuthz(kLogAuthzDebug | kLogAuthzSyslog | kLogAuthzSyslogErr, "Failure getting the ftoken position");
+      return NULL;
   }
 
-  proxy->clear();
-  const unsigned kBufSize = 1024;
-  char buf[kBufSize];
-  unsigned nbytes;
-  do {
-    nbytes = fread(buf, 1, kBufSize, fproxy);
-    if (ferror(fproxy)) {
-      LogAuthz(kLogAuthzDebug | kLogAuthzSyslog | kLogAuthzSyslogErr, "Error reading token file");
+  token->clear();
+  while(true) {
+    int c = fgetc(ftoken);
+    if (c == EOF) {
+      if (ferror(ftoken)) {
+        LogAuthz(kLogAuthzDebug | kLogAuthzSyslog | kLogAuthzSyslogErr, "Error reading token file");
+        return NULL;
+      }
+      break;
     }
-    if (nbytes > 0)
-      proxy->append(string(buf, nbytes));
-  } while (nbytes == kBufSize);
-
-  // Remove the newline at the end of the token
-  if ((*proxy)[proxy->size()-1] == '\n') {
-    proxy->erase(proxy->size()-1);
+    if (c == '\0') {
+      // This will happen when reading from $BEARER_TOKEN environment
+      break;
+    }
+    if (c == '\n') {
+      // This will happen when reading from a file
+      break;
+    }
+    *token += (unsigned char) c;
```

Comment:
I was worried about reading off the end of the environment onto an invalid memory page, although come to think of it that shouldn't happen because it is actually reading from `/proc/$$/environ` and so should hit end of file.  But it does simplify the code quite a bit, fgetc does buffering, I assume the extension of the token string happens in chunks, and this operation doesn't happen very often.